packages/tpetra/core/src/Tpetra_transform_MultiVector.hpp

```diff
@@ -323,6 +323,15 @@ namespace Tpetra {
              range_type range (execSpace, 0, lclNumRows);
 
              Kokkos::parallel_for (kernelLabel, range, g);
+
+             // This fence is currently needed for the following specific situation:
+             // For transform Y to X:
+             //  Y.putScalar()    // acts on device
+             //  Y.sync_host()    // now need_sync_host() and need_sync_device() are false
```

Comment:
Wait, if `putScalar` operates on device data, then `need_sync_host()` should be `true` after calling `putScalar`.  If not, then that's a bug in `putScalar`.

Reply:
@mhoemmen It is need_sync_host() true after the putScalar. But then the user calls sync_host() before calling transform, at which point I verified need_sync_host() and need_sync_device() are now both false, which I think is correct. But then it seems that in this case there is no way for sync state information to provide the fence since Y never actually changes and is already synced to device.

Probably user would normally not do this before the transform and would normally not care to read Y anyways since it's not modified. But not sure how to make sure that one case is logically covered without adding some additional internal logic.

Reply:
@MicheldeMessieres wrote:
> But not sure how to make sure that one case is logically covered without adding some additional internal logic.

It seems like the problem is that `Tpetra::MultiVector::sync_host` doesn't fence.  (This is because it does nothing but call `Kokkos::DualView::sync_host`, and that doesn't fence.)  Why not just have `Tpetra::MultiVector::sync_host` do a fence?

It seems like the goal should be for developers to avoid calling `sync_host`, since that implies they have code that needs to run on host.  Thus, I'm OK with making `sync_host` always fence.  @trilinos/tpetra, what do you all think?